### PR TITLE
[Linux] revise dark theme detection

### PR DESCRIPTION
**Pull request by jpnurmi**
Check whether the window text color is light or dark. This detects dark mode also for other themes than the dark variants of Yaru & Adwaita.

Fixes: flutter/flutter#80199

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

### CODE REVIEW

**Comment by chinmaygarde** [comment]
@cbracken @robert-ancell This looks good to me but I am not familiar with GTK. Can someone take a pass over this review please?

**Comment by jpnurmi** [comment]
If accessing the top-level window to get ahold of the style context is a problem (e.g. in headless mode?), we could revert to checking if the theme name ends with "-dark", which is a pretty common convention for GTK themes.

File: shell/platform/linux/fl_settings_plugin.cc [hunk 1]
```diff
@@ -4,6 +4,8 @@
 
 #include "flutter/shell/platform/linux/fl_settings_plugin.h"
 
+#include <gmodule.h>
+#include <gtk/gtk.h>
```
Comment (cbracken) [inline shell/platform/linux/fl_settings_plugin.cc:8]: nit: add a one-line space between C and C++ includes.

File: shell/platform/linux/fl_settings_plugin.cc [hunk 2]
```diff
@@ -32,6 +33,27 @@ struct _FlSettingsPlugin {
 
 G_DEFINE_TYPE(FlSettingsPlugin, fl_settings_plugin, G_TYPE_OBJECT)
 
+static bool is_dark_color(GdkRGBA* color) {
+  return color->red < 0.5 && color->green < 0.5 && color->blue < 0.5;
+}
```
Comment (cbracken) [inline shell/platform/linux/fl_settings_plugin.cc:38]: We have some very similar logic in the framework. Please take a look:
https://github.com/flutter/flutter/blob/8fe4cc79648a952f9c7e49a5248756c2ff98fa3b/packages/flutter/lib/src/material/theme_data.dart#L1470-L1488

**Review by cbracken — CHANGES_REQUESTED** [review]
Thanks for taking a look at this. Added a couple comments.

File: shell/platform/linux/fl_settings_plugin.cc (hunk 2 above)
Comment (jpnurmi) [reply shell/platform/linux/fl_settings_plugin.cc:65]: Nice, thanks! :+1:

**Comment by cbracken** [comment]
Thanks! Looks good!

/cc @robert-ancell should we keep the "-dark" heuristic, or do you think this is adequate as-is? In particular, @jpnurmi have you verified this works for the adwaita-dark and yaru-dark themes?

**Review by cbracken — APPROVED** [review]

**Comment by jpnurmi** [comment]
> should we keep the "-dark" heuristic, or do you think this is adequate as-is?

I had a "-dark" check as a fallback at first just in case a window wouldn't exist, but then I realized that `fl_engine_start()`, where the settings plugin is created, is called from `fl_view_realize()` after the window has been created.

> In particular, @jpnurmi have you verified this works for the adwaita-dark and yaru-dark themes?

Yes, I have tested the light and dark variants of Adwaita, Yaru, and Materia.

**Comment by cbracken** [comment]
Perfect -- thanks!

**Comment by jpnurmi** [comment]
Oops, done!🙂

**Review by robert-ancell — APPROVED** [review]
It's such a shame we have to jump though these hoops to determine if a theme is dark :(

Changes look fine to me, thanks @jpnurmi!

**Pull request merged**